moped-editor/src/views/projects/projectView/ProjectMilestones.js

```diff
@@ -52,262 +205,219 @@ const ProjectMilestones = ({ projectId, loading, data, refetch }) => {
+      !updatedMilestoneData.description ||
+      updatedMilestoneData.description.trim() === ""
+        ? null
+        : updatedMilestoneData.description;
 
-            // Execute insert mutation
-            return addProjectMilestone({
-              variables: {
-                objects: [newMilestoneObject],
-              },
-            }).then(() => {
-              // Refetch data
-              refetch();
-            });
+    if (updatedRow.isNew) {
+      delete updatedMilestoneData.isNew;
+      delete updatedMilestoneData.id;
+      delete updatedMilestoneData.project_milestone_id;
+
+      return (
+        addProjectMilestone({
+          variables: {
+            objects: {
+              ...updatedMilestoneData,
+              project_id: projectId,
+            },
           },
-          onRowUpdate: (newData, oldData) => {
-            const updatedMilestoneObject = {
-              ...oldData,
-              ...newData,
-            };
+        })
+          .then((response) => {
+            // replace the temporary row id with the id from the record creation
+            const record_id =
+              response.data.insert_moped_proj_milestones.returning[0]
+                .project_milestone_id;
+            updatedRow.project_milestone_id = record_id;
+          })
+          .then(() => refetch())
+          // from the data grid docs:
+          // Please note that the processRowUpdate must return the row object to update the Data Grid internal state.
+          .then(() => updatedRow)
+          .catch((error) => {
+            console.error(error.message);
+          })
+      );
+    } else {
+      // Remove __typename since we removed it from updatedRow and check if the row has changed
+      delete originalRow.__typename;
+      const hasRowChanged = !isEqual(updatedRow, originalRow);
 
-            // Coerce empty strings to null
-            Object.keys(updatedMilestoneObject).forEach((key) => {
-              if (updatedMilestoneObject[key] === "") {
-                updatedMilestoneObject[key] = null;
-              }
-            });
+      if (!hasRowChanged) {
+        return Promise.resolve(updatedRow);
+      } else {
+        return (
+          updateProjectMilestone({
+            variables: updatedMilestoneData,
+          })
+            .then(() => refetch())
+            // from the data grid docs:
+            // Please note that the processRowUpdate must return the row object to update the Data Grid internal state.
+            .then(() => updatedRow)
+            .catch((error) => {
+              console.error(error.message);
+            })
+        );
+      }
+    }
+  };
 
-            // Remove extraneous fields given by MaterialTable that
-            // Hasura doesn't need
-            delete updatedMilestoneObject.tableData;
-            delete updatedMilestoneObject.project_id;
-            delete updatedMilestoneObject.__typename;
+  const handleProcessUpdateError = (error) => {
```

Comment:
I happened to be comparing this to the funding table, and I noticed that we have a snackbar in that table to give user feedback about errors. I'll create a spinoff issue to update all of the tables with snackbars.

Edit: https://github.com/cityofaustin/atd-data-tech/issues/19655